### PR TITLE
Avoid installation caches in image layers

### PR DESCRIPTION
**Pull request by PiDelport**
### Checklist
- [x] I've run `make` from the root directory to see all Dockerfiles are created successfully.
- [x] I've updated the documentation if necessary.

### Motivation and Context

This avoids including the apt-get and pip installation cache files in the Docker image layers, which follows Dockerfile best practices.

This should reduce the size of the Python images for example by about 20 to 30 MB (estimated), which I imagine is significant when multiplied by all pulls.

Background for apt-get:

* https://docs.docker.com/develop/develop-images/dockerfile_best-practices/#apt-get

### CODE REVIEW

**Comment by iynere** [comment]
this looks fantastic tysm @PiDelport 

we're v aware our dockerfiles are highly non-optimal 😩...... we're working on an internal pilot to launch a brand new set of images that are built very differently, which will be super in-line w/ docker best practices, but there's definitely a lot we can cleanup here in the interim 🙏

**Comment by iynere** [comment]
i'll wait for @felicianotech to weigh in but then i'm 👌 to merge

File: android/Dockerfile.m4 [hunk 1]
```diff
@@ -48,15 +51,18 @@ RUN sudo apt-get update && \
     sudo apt-get install --yes \
         xvfb lib32z1 lib32stdc++6 build-essential \
         libcurl4-openssl-dev libglu1-mesa libxi-dev libxmu-dev \
-        libglu1-mesa-dev
+        libglu1-mesa-dev && \
+    sudo rm -rf /var/lib/apt/lists/*
 
 # Install Ruby
-RUN cd /tmp && wget -O ruby-install-0.6.1.tar.gz https://github.com/postmodern/ruby-install/archive/v0.6.1.tar.gz && \
+RUN sudo apt-get update && \
```
Comment (felicianotech) [inline android/Dockerfile.m4]: Why was 'apt-get update` added here?

**Comment by felicianotech** [comment]
This PR looks good to me though I had one inline comment.

@iynere While this PR is technically sound, there is one concern here. In a normal situation, what we should be doing is removing the Apt cache. That's what this PR does, and it makes the images smaller, which is a good thing.

The problem is, because we didn't do this previously, many users can install a package during a build with just `apt-get install` right now and it will work. If we publish this change, all users **will have to** run `apt-get update` before they run `apt-get install` otherwise their build will fail.

So the question is, do we implement a Dockerfile best practice that will make our images slightly more light-weight but will then break some user builds?

If the answer is yes, we should probably at least pre-announce this change in Discuss to give people some heads up.

File: android/Dockerfile.m4 (hunk 1 above)
Comment (PiDelport) [reply android/Dockerfile.m4]: The Ruby install script requires it: it installs Ruby dependencies without running apt-get update first.

See this failure: https://circleci.com/gh/circleci/circleci-images/19631

**Comment by PiDelport** [comment]
@felicianotech: That's a good point.

It should be standard Dockerfile practice to update before installing, but is there any way to gather data on how many builds may not be doing this?

**Comment by iynere** [comment]
@felicianotech yeah, i think we should do it but announce first

i'd say the vast majority of folks know that sometimes you have to `apt-get update` before installing a package—so i think it's worth the tradeoff

**Comment by felicianotech** [comment]
@iynere Want to add this to the mid-2019 milestone? Or should this land sooner?

**Comment by iynere** [comment]
@felicianotech no, i think if we're going to have to announce it in advance, we should add it to the milestone, thank you !

**Comment by felicianotech** [comment]
Hey @PiDelport,

So I've added this PR to our mid-2019 milestone where we plan to merge several changes that will break some user's builds. This allows us, and users, to plan ahead with proper communication around these breaking changes.

So, this will take a couple of months (potentially) to merge, but it will be merged. :)

**Comment by PiDelport** [comment]
@felicianotech: Cool, thanks for the update!

I see this gained a merge conflict; should we try to keep it rebased and up to date until the milestone?

**Comment by Richard-Mathie** [comment]
Just been burnt by this issue. Because the image caches can be out of date the first
`sudo apt update` will fail on building a circle circleci job using an old image.

Please clean up with `sudo rm -rf /var/lib/apt/lists/*` at the end of the image build. It is not super intuitive to work out what/why builds were failing at apt update. It is best practice for more than just image size optimisation.

:+1: to the pull request this is a good idea

**Pull request merged**